### PR TITLE
Translate `_d_arrayassign{,_l,_r}` to templates

### PR DESCRIPTION
**Pull request by teodutu**
- Implement template `_d_arrayassign{_l,_r}`.
- Lower array assignment expressions to the above templates.
- Remove old lowering from e2ir.d.
- Remove the old `_d_arrayassign_{l,r}` and `_d_arrayassig` as its usage was merged with `_d_arrayassign_l`.

### CODE REVIEW

File: compiler/src/dmd/canthrow.d [hunk 1]
```diff
@@ -114,8 +114,10 @@ extern (C++) /* CT */ BE canThrow(Expression e, FuncDeclaration func, bool mustN
                         import dmd.id : Id;
 
                         auto sd = ts.sym;
+                        auto id = ce.f.ident;
```
Comment (thewilsonator) [inline compiler/src/dmd/canthrow.d:117]: this could be const

File: compiler/src/dmd/expressionsem.d [hunk 2]
```diff
@@ -9969,10 +9969,89 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+        if (t1b.ty != Tsarray && t1b.ty != Tarray)
+            return ae;
+
+        const isArrayAssign =
+            (ae.e1.isSliceExp || ae.e1.type.ty == Tsarray) &&
+            (ae.e2.type.ty == Tsarray || ae.e2.type.ty == Tarray) &&
+            (ae.e1.type.nextOf && ae.e2.type.nextOf && ae.e1.type.nextOf.mutableOf.equals(ae.e2.type.nextOf.mutableOf));
+
+        if (isArrayAssign)
+        {
```
Comment (thewilsonator) [inline compiler/src/dmd/expressionsem.d:10008]: use early return here
```d
if (!isArrayAssign)
    return ae;
...
```

**Review by thewilsonator — APPROVED** [review]

**Comment by thewilsonator** [comment]
Could you please double check that the build kite failures are unrelated? Otherwise this looks good. Please ping me when you're done.

**Review by RazvanN7 — APPROVED** [review]

**Comment by teodutu** [comment]
@thewilsonator The buildkite failures were related. They were caused by the temporary variable used for `_d_arrayappend_r`:
```d
_tmp = rhs, _d_arrayappend_r(lhs, _tmp);
```

I was not adding the `maybescope` storage class to `_tmp` so the compiler believed a `scope` rhs was excaping. This caused both buildkite failures. They should work fine now.

File: compiler/src/dmd/dinterpret.d [hunk 3]
```diff
@@ -4831,27 +4831,32 @@ public:
...
             {
-                // In expressionsem.d `T[x] ea = eb;` was lowered to `_d_array{,set}ctor(ea[], eb[]);`.
+                // In expressionsem.d:
+                // * `T[x] ea = eb;` was lowered to `_d_array{,set}ctor(ea[], eb[]);`.
+                // * `ea = eb` (ea and eb are arrays) was lowered to `_d_arrayassign_l(ea[], eb[])`.
                 // The following code will rewrite it back to `ea = eb` and then interpret that expression.
-                if (fd.ident == Id._d_arraysetctor)
-                    assert(e.arguments.dim == 2);
-                else
-                    assert(e.arguments.dim == 3);
```
[the remark this answers is not in the document]
Comment (teodutu) [reply compiler/src/dmd/dinterpret.d:4841]: By mistake. I put them bak.

**Review by RazvanN7 — APPROVED** [review]
LGTM!

**Comment by RazvanN7** [comment]
cc @dkorpel

**Comment by schveiguy** [comment]
This change introduced unnecessary out-of-bounds checking when doing array operations. e.g.:  https://github.com/dlang/dmd/pull/14310/files#diff-c9a721f453e1e17d997f8603ca3e4e65c765f0ebf0dd1cfb972ddddc959bdbb1R92-R94

**Comment by teodutu** [comment]
@schveiguy Are these checks relevant given that they're disabled when compiling with maximum optimisations, i.e. ` -O -release -inline -boundscheck=off`?

**Comment by schveiguy** [comment]
Yes, they are relevant. First, because they are redundant, the entire operation is bounds checked first (see [here](https://github.com/dlang/dmd/pull/14310/files#diff-c9a721f453e1e17d997f8603ca3e4e65c765f0ebf0dd1cfb972ddddc959bdbb1R49))

And second, because it's a template, it will be compiled under whatever options the user is using at the time.

**Comment by CyberShadow** [comment]
This pull request introduced a regression:
https://issues.dlang.org/show_bug.cgi?id=23959

**Pull request merged**